### PR TITLE
Float.is_integer updated for removing disparity in gamma

### LINKED ISSUE #16847: Disparity in gamma function result for non-positive float numbers
As nonpositive integers are poles for gamma function, we have the following:
```
>>> gamma(-1)
zoo
>>> gamma(0)
zoo
```
But for nonpositive integers represented as floats, we get the following error
```
>>> gamma(-1.0)
Traceback (most recent call last):
  File "/home/ritesh/Desktop/sympy/sympy/core/evalf.py", line 1308, in evalf
    rf = evalf_table[x.func]
KeyError: gamma

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ritesh/Desktop/sympy/sympy/core/cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/home/ritesh/Desktop/sympy/sympy/core/function.py", line 463, in __new__
    result = result.evalf(mlib.libmpf.prec_to_dps(pr))
  File "/home/ritesh/Desktop/sympy/sympy/core/evalf.py", line 1442, in evalf
    result = evalf(self, prec + 4, options)
  File "/home/ritesh/Desktop/sympy/sympy/core/evalf.py", line 1314, in evalf
    xe = x._eval_evalf(prec)
  File "/home/ritesh/Desktop/sympy/sympy/core/function.py", line 586, in _eval_evalf
    v = func(*args)
  File "/home/ritesh/venv/lib/python3.6/site-packages/mpmath/ctx_mp_python.py", line 1012, in f
    return ctx.make_mpf(mpf_f(x._mpf_, prec, rounding))
  File "/home/ritesh/venv/lib/python3.6/site-packages/mpmath/libmp/gammazeta.py", line 1933, in mpf_gamma
    raise ValueError("gamma function pole")
ValueError: gamma function pole
```
I think there should be a check like `floor(x) == x` and then we can return `gamma(floor(x))`, this will help us to avoid the `ValueError` as:
```
>>> gamma(floor(-1.0))
zoo
```

### PR DESCRIPTION
**Pull request by czgdp1807**
#### References to other Issues or PRs

[1] Fixes https://github.com/sympy/sympy/issues/16847
[2] https://github.com/sympy/sympy/issues/16847#issuecomment-493653572

#### Brief description of what is fixed or changed
Float.is_integer now returns `True` for values like, `-1.0`,  which helps in removing disparity as mentioned in the issue referenced above. 

#### Other comments
ping @smichr @jksuom @ritesh99rakesh 

The following test is failing due to the changes,
```
__________ sympy/core/tests/test_power.py:test_issue_6100_12942_4473 ___________
Traceback (most recent call last):
  File "/home/gagandeep/sympy/sympy/core/tests/test_power.py", line 284, in test_issue_6100_12942_4473
    assert ((x*y)**1.0).func is Pow
AssertionError
```
`(x*y)**1.0` is converted to `x**1.0*y**1.0` returning `Mul` instead of `Pow` therefore failing the test.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/core/numbers.py [hunk 1]
```diff
@@ -1254,7 +1254,7 @@ def _eval_is_infinite(self):
         return False
 
     def _eval_is_integer(self):
-        return self._mpf_ == fzero
+        return self._mpf_[2] == fzero[2]
```
Comment (jksuom) [inline sympy/core/numbers.py]: This will only work with odd integers. For all integers, we have `self._mpf_[2] >= 0`.

**Comment by smichr** [comment]
I don't think you want to change Float(1).is_integer to be True. The lower-case variety means that you have a symbolic integer (as opposed to a literal Integer). 

Floats are a special subset of Rationals. Although there is no inherent uncertainty in the representation of a given Float, each float must represent numbers which are too close to it to be represented with finite precision. So even Float(1) -- though representationally equal to 1 -- stands for 1 +/- eps and should not be used to separate powers as in `(x*y)**1.0` to `x**1.0*y**1.0`. 

But what about `gamma(Float(-1))`? Even though we might say that `Float(-1)` does not exactly stand for -1, it *is* -1: all numbers in `[-1-eps, -1+eps]` are being represented as -1 (like numbers underflowing to 0) and we can't compute gamma at -1. 

So either Float(1).is_integer becomes True and all tests for foo.is_integer become foo.is_integer and not foo.is_Float -- that's not a good solution.  Or, for places where you are looking for a Number that is representationally the same as an Integer, use `as_int(foo, strict=False)`:
```python
>>> if as_int(Float(1), strict=False): print('ok')
...
ok
```

File: sympy/core/numbers.py [hunk 2]
```diff
@@ -1254,7 +1254,7 @@ def _eval_is_infinite(self):
         return False
 
     def _eval_is_integer(self):
-        return self._mpf_ == fzero
+        return self._mpf_[2] >= 0
```
Comment (smichr) [inline sympy/core/numbers.py]:
```suggestion
        return self._mpf_ == fzero
```

File: sympy/functions/special/gamma_functions.py [hunk 3]
```diff
@@ -107,7 +107,7 @@ def eval(cls, arg):
                 return S.NaN
             elif arg is S.Infinity:
                 return S.Infinity
-            elif arg.is_Integer:
+            elif arg.is_integer:
```
Comment (smichr) [inline sympy/functions/special/gamma_functions.py]: at the top of file

from sympy.core.compatibility import as_int

then here,

```python
elif as_int(arg, strict=False):
```

**Comment by smichr** [comment]
p.s. sorry about the typo that suggested this PR would close because of the one I was working on!

**Comment by jksuom** [comment]
> all tests for foo.is_integer become foo.is_integer and not foo.is_Float

I'm not sure I understand why this would be necessary. We have `.is_Integer` for those cases where an Integer object is needed. But otherwise integer valued floats will compare equal to proper integers so it looks like they should also be accepted as 'integers'. (But maybe I don't see the whole picture.)

**Comment by czgdp1807** [comment]
> But otherwise integer valued floats will compare equal to proper integers so it looks like they should also be accepted as 'integers'. (But maybe I don't see the whole picture.)

I agree with @smichr here. Float(1.0) and 1 are different as in 1.0 precision matters but 1 is a constant with infinite precision. AFAIK, there is always an error of the max precision. If I am wrong then please correct me, at least my concepts about floats and integers in general would be corrected. For now, I will make the change as @smichr suggested. Thanks.

**Comment by czgdp1807** [comment]
@smichr Tests passing. Is it good to merge?

**Comment by smichr** [comment]
> looks like they should also be accepted as 'integers'

This is like the [face/vase optical illusion](https://hubpages.com/art/Two-Faces-or-a-Vase-10-Simple-but-Wonderful-Optical-Illusions):
image

A Float is a subset of Rationals and is exact. When used in a calculation it will give an approximate result. Even though all numbers `[f-eps, f+eps]` are mapped to f, that is our problem, not `f`'s problem. So when trying to decide if `(x*y)**1.0` should be expanded as `x**1.0*y**1.0` we must choose to do so. Refusingto do so because there are other numbers that *could* be 1.0 is the "face"; to do so because the 1.0 that we have is identically 1 is the "vase". 

To me it seem like the result that we *calculate* could be exact, but currently that is not how SymPy behaves:
```python
>>> pi.n()**Float(1,2) == pi.n()**Float(1)
True
```
Whether that should be False is not lilkely to change since
```python
>>> mlib.mpf_mul(Float(.1)._mpf_,Float(1,2)._mpf_)
(0, 3602879701896397, -55, 52)
>>> mlib.mpf_mul(Float(.1)._mpf_,Float(1,22)._mpf_)
(0, 3602879701896397, -55, 52)
```
So why not replace input to Float that has a power-of-2 denominator with a Rational? That's another issue. Presumably this would not be a trouble with code generation because each language printer would have to figure out if 1/2 should be printed like that or as 1/2.0 given the context.

I am in favor of the previous modifications and an update to the test to show `(x*y)**1.0` -> `x**1.0*y**1.0`.

**Comment by smichr** [comment]
The details of handling Floats as ints/rationals may take a while to work out and this patch is a reasonable interim solution. I added a test that directs the author to remove `intlike` if `Float(1).is_integer` ever becomes True.

**Comment by oscarbenjamin** [comment]
I don't see what's gained by preventing `(x*y)**1.0 -> x**1.0*y**1.0` but I agree that this PR shouldn't be held up over that.

We really need to get a clear vision for what to do with floats but it's difficult to get there just by sporadic discussion on myriad unrelated issues

**Comment by czgdp1807** [comment]
@oscarbenjamin @smichr I opened this PR for correcting the disparity in gamma function and not floats. As far as I can figure out that is solved by @smichr 's suggestion and that too without affecting the `Float` which I think is quite acceptable.
For discussing over `Float` we can have an issue and a separate PR fixing that.
I will think upon the recent comments made after the last commit and will write my view in the comments.

**Comment by czgdp1807** [comment]
Thanks @smichr. I will shortly raise an issue for discussing over `Float`, especially the `is_integer` attribute.
However, if there is any such pre-existing method, then please ping me there. I want to be a part of that discussion. Thanks. :)

**Comment by jksuom** [comment]
Changing `Float.is_integer` is probably not good but it would be possible to use the exponent criterion directly: `arg.is_Float and arg.num.exp >= 0`.

**Pull request merged**